1168 - Upload videos to Kaltura and add Kaltura player

zifgu opened the pull request:
## Description
In this PR we want to add functionality to upload videos from our local server to Kaltura, to reduce load and storage use on our servers. 

Users will still use the regular file upload to Wordpress to add videos. Admin users can use a dashboard in the Settings modal to trigger an upload of local videos in the Tapestry to Kaltura. The transfer happens in the backend, and when done, the video data in the nodes will be silently changed to the video data from Kaltura.

![image](https://user-images.githubusercontent.com/44929104/178617526-77cefb8c-e853-4d1c-81f1-9108ef83b693.png)

We also add a Kaltura media player to play Kaltura videos. Users can use the player by uploading videos to Kaltura and telling it to switch uploaded videos to the Kaltura media player, or by directly specifying the Kaltura ID of the video.

## Changes
### Video upload
- Adds endpoint to upload videos to Kaltura using the Kaltura API (only accessible by users with "administrator" role). Videos are uploaded in batches of 10.
- Only 1 upload can be ongoing at a time. If a Kaltura upload is already in progress, the endpoint will do nothing.
- Uploaded videos are organized using [Kaltura Categories](https://developer.kaltura.com/api-docs/Secure_Control_and_Govern/Content-Categories-Management.html), which have a tree-like hierarchy. Videos are saved under the category hierarchy `Tapestry > {site URL} > {date OR Tapestry name}`.

Frontend:
- Re-fetches the video URL stored in the node if the existing URL becomes unavailable while the frontend is open (for example, if the video was uploaded and deleted from local server)

### Tracking uploads

- Adds Kaltura Sync operation modal to start uploads and track upload progress.
    - The modal should only be accessible if Kaltura is available on the backend and user has the "administrator" role.
    - Users can select the videos they wish to upload.
    - Users have an option to switch uploaded videos to open with the Kaltura media player.
    - Users can click a button to stop the upload prematurely. (Since videos are uploaded in batches, the stop signal will prevent future batches from uploading, but still wait for the current batch to finish.)
    - Upload progress includes upload status (e.g. not started, uploading, error, etc), Kaltura ID, and additional information (e.g. error messages) for each video.
        - If no upload is in progress, displays the statuses from the most recent upload.
        - Only shows videos in the current Tapestry, to avoid confusion.
    - Upload progress refreshes automatically every 15 seconds. Users can also refresh manually.

Changes on the Tapestry Plugin Settings admin page:
- Restructures the page to support more than 1 settings section
- Changes the "Clean h5p nodes" setting from a checkbox to a button
- Adds setting that allows users to choose whether to categorize videos by date (`Tapestry > {site URL} > {date}`) or Tapestry name (`Tapestry > {site URL} > {Tapestry name}`). Default is by date.
- Adds a button that finds any nodes marked "Converting" and refreshes their upload status. This is to clean up the upload status in the wake of an interrupted upload or error.
- Adds a button to forcefully mark the upload as not in progress – intended as a last resort for fixing upload issues.

### Kaltura Media Format

- Adds a select field to the ContentForm when media type is "video". User must now explicitly choose the media format out of URL, Youtube, and Kaltura. Defaults to URL.
- User can create a Kaltura video by entering a Kaltura video ID or uploading a video directly to Kaltura.
- User can choose the regular player (UrlVideoMedia) _or_ Kaltura player for Kaltura videos. If Kaltura player is chosen but `partnerId`, `serviceUrl`, or `uniqueConfiguration` are not found in the node's `typeData.kalturaData`, falls back to regular player.
- Adds Kaltura video player, similar to other video media components. Specifically, it should work with Popups and as a sub-item of a Multicontent node.

### Other
- Adds endpoints to retrieve information about a Kaltura video, given the Kaltura ID
- Adds a section to the README about setting up Kaltura
- Fixes detection of TAPESTRY_USE_DEV_MODE
- Defines TAPESTRY_USE_DEV_MODE and loads Kaltura configuration secrets into the wp-config file during Wordpress setup for CI
- Adds a step to the CI workflow that installs Composer dependencies
- Uses Bootstrap's `<b-badge>` component for displaying notifications on the toolbar buttons, including the review comments button and the operations button.

## Screenshot

Upload dashboard:

https://user-images.githubusercontent.com/44929104/186055625-ba72a650-2a76-4d6b-9f4a-5ca5f3a1b672.mp4

Changes to Tapestry Plugin Settings page:
![image](https://user-images.githubusercontent.com/44929104/176565294-5848489a-b042-4265-a936-6932289de90f.png)

Kaltura player:

https://user-images.githubusercontent.com/44929104/187575585-4098cc65-5495-4d55-a837-bb237e5135ca.mp4

## Testing
Before testing the PR for the first time, first make sure to follow the README to ensure Kaltura works.

## Automated Testing
- Tests creating a Kaltura video node using a valid Kaltura ID, and that the video opens with the Kaltura player.

## Helpful Links
Uploading videos to Kaltura: https://developer.kaltura.com/workflows/Ingest_and_Upload_Media/Uploading_Media_Files

Kaltura Categories: https://developer.kaltura.com/api-docs/service/category

Kaltura player API: http://player.kaltura.com/docs/api

## Issue Linkage
Closes #1051
Closes #1109 
Closes #1168

legendword reviewed (changes requested):
The dashboard works as intended, video uploads are triggered successfully and the status refreshed automatically. I have left some comments for code changes but other than those, everything looks good to me.

zifgu commented on settings.php:
```diff
@@ -13,62 +14,108 @@ function tapestry_settings_init()
+
+function load_tapestry_settings_page_scripts($hook)
+{
+    // TO DO: find a better way to test the current settings page
```
My bad, I forgot I left a comment in here! That reasoning makes sense. I'll add a comment explaining the decision to stick to checking the suffix.

zifgu replied:
Never mind, I found out a better way to check this. The function that creates the sub-page for Tapestry settings actually returns the full slug of the new page, and we can just pass it to `load_tapestry_settings_page_scripts` to check against the `hook` argument.

legendword replied:
That's great!

zifgu commented on endpoints.php:
```diff
@@ -340,6 +364,21 @@ function () use ($ENDPOINT, $REST_API_NAMESPACE) {
+
+function getKalturaUploadStatus($request)
+{
+    return get_option('tapestry_kaltura_upload_log');
```
Thanks for catching this. I have implemented the suggestion.

zifgu commented on endpoints.php:
```diff
@@ -322,6 +322,30 @@
+    'UPLOAD_VIDEOS_TO_KALTURA' => (object) [
+        'ROUTE' => '/kaltura/upload_videos',
+        'ARGUMENTS' => [
+            'methods' => $REST_API_GET_METHOD,
```
You're right, it looks like POST is more suited for actions that have side effects or create objects. I think we can change it to POST.

legendword reviewed (approved):
The changes look good to me.

legendword reviewed (approved):
The upload functionality works as intended; stopping the upload successfully stops after current batch is complete. Looks good to me.

zifgu commented:
A few more improvements to make:
- Currently, we are checking local uploads by checking if the mediaURL contains `/wp-content/uploads/`, but this could include some URLs from WordPress sites that are not our own. The check should be fixed to include only the local upload directory.
- If the Kaltura upload is interrupted unexpectedly, videos in the middle of being uploaded will need to be re-uploaded.
    - To avoid having to re-upload them, keep track of the upload status in the `typeData`.
    - Add a function to the WP Admin dashboard that updates the status of any videos that are still marked "converting".

zifgu commented:
Based on testing, I think the current upload code may be using the wrong status type.

The [Kaltura documentation for `media->get`](https://developer.kaltura.com/api-docs/service/media/action/get) suggests that the `status` we are checking is not of type [FlavorAssetStatus](https://developer.kaltura.com/api-docs/General_Objects/Enums/KalturaFlavorAssetStatus) but [EntryStatus](https://developer.kaltura.com/api-docs/General_Objects/Enums/KalturaEntryStatus).

Flavor assets seem to be a different type of object in Kaltura, associated with particular videos, not the video itself.

Results from logging/inspecting the response and status support this. The response is of type Kaltura\Client\Type\MediaEntry, and its `status` property is a string (like EntryStatus), not an integer (like FlavorAssetStatus).

- After step 3 of the upload, before the video file is attached to the Kaltura media entry, the `status` is `"7"` which would be `EntryStatus::NO_CONTENT`.
- After the video is uploaded, while it is converting, the status is `"1"` which would be `EntryStatus::PRECONVERT`.
- If there is an error converting (for example, if the video file is corrupted), the status becomes `"-1"`, which would be `EntryStatus::ERROR_CONVERTING`.

It seems likely to me that the `status` is of type EntryStatus, so I am implementing checks for specific statuses against the different possible values of EntryStatus.

zifgu commented:
@legendword It would be great if you could review the changes described in this comment: https://github.com/tapestry-tool/tapestry-wp/pull/1169#issuecomment-1167839498. 

The base upload code has been refactored a bit, so I would also appreciate if you tested that it still works on your end. Since the `kalturaData` field is not displayed, unfortunately, you will have to check that it is updated through the dev tools.

legendword commented on utilities/class.tapestry-helpers.php:
```diff
@@ -307,4 +307,125 @@ public static function nodeNeighbourIsPublished($nodeMetaId, $tapestryPostId)
+
+        if ($kalturaData) {
+            $typeData->kalturaData['id'] = $kalturaData->id;
+            $typeData->kalturaData['partnerId'] = $kalturaData->partnerId;
```
Is this particular `partnerId` property being used? I did a search and couldn't find any other occurrences.

zifgu replied:
Yes, I think it is not used anywhere. It was saved in the original upload code, though I am not sure why.

Since it's not being used, I will just delete it for now and it should be easy to add back in the future.

legendword commented on templates/vue/src/components/modals/SettingsModal/KalturaUploadTab.vue:
```diff
@@ -0,0 +1,217 @@
+      this.uploadStatusRefreshTimer = 0
+    },
+    startVideoUpload() {
+      console.log("Starting upload")
```
```suggestion
```

zifgu replied:
Oops! Thank you.

legendword reviewed (changes requested):
Tested uploading a video to Kaltura using the Kaltura tab in the Tapestry Settings modal. The node's typeData.kalturaData is being updated to the appropriate values in every step of the process, and the existing features continue to function as desired.

When commenting out the setting of video uploadStatus in the backend to mimic potential problems with PHP timeout or server shutdown preventing the update of uploadStatus, the video will appear as Converting in the Kaltura tab of the Tapestry Settings; after pressing the Clean Uploaded Videos button in the Wordpress Dashboard Tapestry Settings page, the video's status is updated to "Complete" and everything else is updated correctly.

I have left some code refactoring suggestions; they are mostly minor.

We should discuss about what to do when the process times out. Either we can notify the user (if the user still has the Tapestry page initiating the upload open) of the time out error, or we can use PHP's set_time_limit to extend the execution time and avoid the time out error.

There is one thing I noticed: the alert for upload in progress in the Wordpress Dashboard Settings page is shown above the Database Settings section which may be confusing to the user (see below).

<img width="985" alt="image" src="https://user-images.githubusercontent.com/12668055/176325994-885607c5-3f77-41b2-897b-acad4bfe05e8.png">

zifgu commented on tapestry.php:
```diff
@@ -349,83 +352,218 @@ function tapestry_tool_log_event()
+            } catch (Exception $e) {
+                $error_start = 'Unable to upload video "'.$video->file->name.'" to Kaltura due to: \n';
+
+                error_log($error_start.$e);
```
Using the error object `$e` logs both the message and the stack trace, which I thought would be more informative. In the latest change, I have made the code's intention a bit clearer.

(Also, as an aside, although we should use single quotes, I can only print newlines to the error_log using double quotes. This would be why the new code uses double quotes.)

zifgu commented on endpoints.php:
```diff
@@ -1550,10 +1592,188 @@ function getQuestionHasAnswers($request)
+
+            if (isset($kalturaData)
+                && is_array($kalturaData)
+                && array_key_exists('uploadStatus', $kalturaData)
```
That makes sense. This suggestion has been added.

zifgu commented:
> There is one thing I noticed: the alert for upload in progress in the Wordpress Dashboard Settings page is shown above the Database Settings section which may be confusing to the user (see below).

Unfortunately, if we use the WordPress admin notice style, the notice will always be at the top of the page. According to [this answer](https://wordpress.stackexchange.com/a/220652), WordPress styles place it under the first h2 (the page title).

I agree that this is not the ideal behaviour, so I have changed it to a custom info box. This is slightly less future-proof since it relies on the contents of some WordPress styles, but the amount of customization is minimal so far.

![image](https://user-images.githubusercontent.com/44929104/176341286-a123e32e-94d1-4dd8-9fcb-2b1871ca5021.png)

zifgu commented on tapestry.php:
```diff
@@ -349,83 +352,218 @@ function tapestry_tool_log_event()
+ *
+ * @param object $upload_request    HTTP request body
+ */
+function upload_videos_to_kaltura($upload_request)
```
This makes sense to me – now that the upload takes parameters for the videos to upload, we probably do not want to make it a cron job. I've moved the code to `endpoints.php`.

zifgu commented:
> Either we can notify the user (if the user still has the Tapestry page initiating the upload open) of the time out error, or we can use PHP's set_time_limit to extend the execution time and avoid the time out error.

I have implemented a handler that marks the upload as no longer in progress after a fatal error and alerts the user to the presence of an error. No matter how much we extend the time limit (or for another reason), a fatal error could still occur, so I think it would be good to reset the upload.

legendword commented on endpoints.php:
```diff
@@ -1630,6 +1638,9 @@ function uploadVideosToKaltura($request)
         update_option(KalturaUpload::STOP_UPLOAD_OPTION, KalturaUpload::NO_VALUE, false);
+        update_option(KalturaUpload::UPLOAD_ERROR_OPTION, '');
+
+        add_action('shutdown', 'cleanUpKalturaUpload');
```
Does this add the callback to the hook every time this uploadVideosToKaltura function is run? If so, it seems like initiating multiple uploads may result in several callbacks being registered. Although, I haven't confirmed if this is actually true.

legendword replied:
After reading about the PHP [register_shutdown_function](https://www.php.net/manual/en/function.register-shutdown-function.php), it seems like the registered callback is only for the current process? If so, then I think this works fine.

zifgu replied:
Yes, I think it is only for the current process. I tested it with some print statements and the shutdown hook was only called when the upload endpoint was called.

legendword reviewed (approved):
The changes all make sense to me.

wynnset commented:
I'm going to merge all the Kaltura PRs into one so it's easier to review and test them.

zifgu commented:
**Changes to test since last review:**

KalturaModal / upload dashboard:
- [ ] Should be able to see uploadable videos, and upload videos successfully, on sites with month- and year-based uploads.
- [ ] All checkboxes should have accessible labels.
- [ ] No unexpected behaviour when reopening and closing the Kaltura modal. (Previously, the `selectedVideos` would be preserved and cause the "select all" checkbox to be checked even when no videos were selected.)
- [ ] When an error occurs during the upload, its stack trace should not be shown.
- [ ] The upload error alert should be dismissible.

NodeModal:
- [ ] When "Use Kaltura Player" is toggled, the media URL and Kaltura ID should be preserved.

Kaltura player:
- [ ] A Kaltura video should be playable when Kaltura is not available, whether it was added via upload or via the NodeModal. 
    - [ ] This is because the following fields are stored under `typeData.kalturaData`. If one or more field is not available, it should fall back to the URL media player.
![image](https://user-images.githubusercontent.com/44929104/186749548-1f9a7c18-ffa0-4953-870d-b04a0418084f.png)


Tapestry Settings page:
- [ ] "Clean Uploaded Videos" should work properly when videos are found to be still converting. (Previously, it incorrectly threw an "Expected the video to be converting, but it was not" error)
- [ ] "Clean Uploaded Videos" should make the upload error disappear.

legendword reviewed (approved):
Tested all new features as outlined in https://github.com/tapestry-tool/tapestry-wp/pull/1169#issuecomment-1227662030. They are all working as expected. The code looks really good, too.

zifgu commented on templates/vue/src/components/Toolbar/TapestryOperationsButton.vue:
```diff
@@ -31,18 +42,27 @@
+import KalturaModal from "@/components/modals/KalturaModal"
 import OtherOperationsModal from "@/components/modals/OtherOperationsModal"
 import { names } from "@/config/routes"
+import { getKalturaStatus } from "@/services/wp"
```
That's true. For consistency, I'll change it to `import * as wp from "@/services/wp"`.

zifgu commented on classes/class.kaltura-api.php:
```diff
@@ -37,7 +39,7 @@ public function getKClient($type = SessionType::USER)
 
             if (!isset($ksession)) {
-                throw new TapestryError('Unable to establish Kaltura session.');
+                throw new TapestryError('FAILED_TO_ESTABLISH_KALTURA_SESSION');
```
I see, that's a good observation. Right now, we do add `catch (TapestryError $e)` around all KalturaApi method calls in #1239, which will handle these errors once it is merged. If we plan on using this PR without #1239 for some time, I can add the try-catch here.

zifgu commented:
**Changes since last review:**
- [ ] Upload error is only shown for the Tapestry in which it occurs.
- [ ] If videos are uploading from this Tapestry, the "Stop Upload" button should be interactable and the "Start Upload" button should show a spinner.
- [ ] If an upload is in progress from a different Tapestry, the user should see a message indicating this. The "Stop Upload" and "Start Upload" buttons should both simply be disabled.

legendword reviewed (approved):
Can confirm that the upload table and buttons are disabled when in a Tapestry different than the one that initiated the upload, and the appropriate message is given to the user:

<img width="798" alt="image" src="https://user-images.githubusercontent.com/12668055/187003402-db6c1f83-dc9b-4d94-b3ca-49c71041a16a.png">

An error during the upload only appears in the tapestry that uploaded it. This is a frontend-only change, meaning the error is still given to the Kaltura modal of all the Tapestries, except it will only be shown in the Tapestry that initiated the latest upload. I think this is reasonable and the code changes are pretty straightforward.

zifgu commented:
Note: ignore the `(Note: this may happen multiple times, so the NodeModal has been changed to no longer add duplicate errors)` in the last commit message. This change was not necessary for this commit.

legendword commented:
Confirmed that successfully uploading a video from the node modal results in the following message:

<img width="792" alt="image" src="https://user-images.githubusercontent.com/12668055/189267833-d82db89a-81ae-4069-a94b-7943c6e8c982.png">

... and that any uploads, whether succeeded or not, made in the node modal will not be recorded in the Kaltura upload log in the Kaltura modal.

If user accidentally closes the node modal by clicking outside of the modal during an upload to Kaltura, the upload will still complete but the node's data will not be updated to reflect the uploaded video. We currently only disable the submitting of the node through the node modal footer while an upload is happening, but do nothing in terms of preventing the user from closing the node modal or closing the browser tab.

Tested uploading a node with a video URL that points to a nonexistent file using the Kaltura modal. The upload completes almost instantly and no feedback is shown in the Upload tab. I can only figure out what happened by switching to the Log tab and viewing the status of the latest upload, which indicates the error clearly:

<img width="765" alt="image" src="https://user-images.githubusercontent.com/12668055/189268566-6f1dad9e-65a9-409e-af2c-3058e15763dc.png">

... when a file is not found, it is not an error from Kaltura's side, and thus no error message will be shown in the Upload tab. Just describing this behaviour in case we want to also inform the user that one or more videos failed to upload due to this particular error (file not found).

Tested overriding Kaltura configuration, and it is still working as expected.

I made some small changes which I will note below:

[x] Include leading zero in the time format for upload log.
[x] Move KalturaModal, KalturaUploadTab, KalturaUploadLogTab into a subdirectory called KalturaModal since they are related components.
[x] Add notice when upload is in progress for the Kaltura modal, prompting the user to view the Log tab if they want to see the status of the upload.
[x] Clear the upload success alert text when initiating a new upload in the node modal, if the user chooses to upload another video to Kaltura right after they have just uploaded one.

legendword commented:
Users will now see a red dot when a Kaltura upload has completed. This requires a change to the refresh timer so that it will keep refreshing every 15 seconds when the Kaltura modal is closed (but will only do so when an upload is in progress).

<img width="415" alt="image" src="https://user-images.githubusercontent.com/12668055/191595424-00647b71-20db-4f79-8f19-857999f46c60.png">

<img width="233" alt="image" src="https://user-images.githubusercontent.com/12668055/191595456-b7c9a82a-1ae1-481d-bc2d-15d76aa030d7.png">

legendword commented:
Now showing number of videos uploaded after last upload completes:
<img width="789" alt="image" src="https://user-images.githubusercontent.com/12668055/192079850-cec588a7-8144-41b6-8777-1bb0e977cc4f.png">

...or number of failed videos if not all videos were uploaded successfully:
<img width="798" alt="image" src="https://user-images.githubusercontent.com/12668055/192081236-c236f551-6308-4525-9008-765fc69ccd8d.png">

New, pill-shaped Bootstrap badge component for the notification count of both the review comments button and the operations button:
<img width="418" alt="image" src="https://user-images.githubusercontent.com/12668055/192080828-65149738-583b-4790-b73b-b5d78793fb6b.png">

The "Upload to Kaltura" menu item also shows the same notification badge:
<img width="236" alt="image" src="https://user-images.githubusercontent.com/12668055/192080877-222940ee-8463-44df-a4a4-9612a4396fd3.png">

The number on the badge for Kaltura is the number of videos uploaded (both successfully and unsuccessfully) for the latest upload in this tapestry.

The notification badge will not disappear unless the Kaltura modal is opened. It will not disappear even if the page is refreshed.

legendword reviewed (approved):
The code refactoring looks great! The Kaltura functionalities are still working as expected.
